File: leetcode_cpp/124.Binary_Tree_Maximum_Path_Sum.cpp
// https://leetcode.com/problems/binary-tree-maximum-path-sum/

class Solution {
 private:
  int res = INT_MIN;

 public:
  int maxPathSum(TreeNode* root) {
    dfs(root);
    return res;
  }

  int dfs(TreeNode* node) {
    if (node == NULL) return 0;

    int left = dfs(node->left);
    int right = dfs(node->right);

    res = max(res, node->val + left + right);
    return max(0, node->val + max(left, right));
  }
};

File: leetcode_java/124.Binary_Tree_Maximum_Path_Sum.java
// https://leetcode.com/problems/binary-tree-maximum-path-sum/

class Solution {
  private int max = Integer.MIN_VALUE;

  public int maxPathSum(TreeNode root) {
    dfs(root);
    return max;
  }

  private int dfs(TreeNode root) {
    if (root == null) {
      return 0;
    }

    int left = dfs(root.left);
    int right = dfs(root.right);

    max = Math.max(max, root.val + left + right);
    int cur = Math.max(0, root.val + Math.max(left, right));

    return cur;
  }
}